src/infra/sslQueryRepo.go

```diff
@@ -0,0 +1,181 @@
+package infra
+
+import (
+	"errors"
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/speedianet/sam/src/domain/entity"
+	"github.com/speedianet/sam/src/domain/valueObject"
+	infraHelper "github.com/speedianet/sam/src/infra/helper"
+)
+
+type SslQueryRepo struct {
+	olsHttpdConfigPath string
+}
+
+type HttpdVhostConfig struct {
+	VirtualHost string
+	FilePath    string
+}
+
+func NewSslQueryRepo() *SslQueryRepo {
+	return &SslQueryRepo{
+		olsHttpdConfigPath: "/usr/local/lsws/conf/httpd_config.conf",
+	}
+}
+
+func (repo SslQueryRepo) SslFactory(
+	sslId int,
+	sslHostname string,
+	sslPrivateKey string,
+	sslCertContent string,
+) (entity.Ssl, error) {
+	var ssl entity.Ssl
+	id, err := valueObject.NewSslId(sslId)
+	if err != nil {
+		return ssl, errors.New("SslIdError")
+	}
+
+	privateKey, err := valueObject.NewSslPrivateKey(sslPrivateKey)
+	if err != nil {
+		return ssl, errors.New("SslPrivateKeyError")
+	}
+
+	var certificate valueObject.SslCertificate
+	var chainCertificates []valueObject.SslCertificate
+	sslCertContentSlice := strings.SplitAfter(sslCertContent, "-----END CERTIFICATE-----\n")
+	for sslCertContentIndex, sslCertContentStr := range sslCertContentSlice {
+		if sslCertContentIndex == 0 {
+			certificate, err = valueObject.NewSslCertificate(sslCertContentStr)
+			if err != nil {
+				return ssl, errors.New("SslCertificateError")
+			}
+			continue
+		}
+
+		chainCertificate, err := valueObject.NewSslCertificate(sslCertContentStr)
+		if err != nil {
+			return ssl, errors.New("SslCertificateError")
+		}
+		chainCertificates = append(chainCertificates, chainCertificate)
+	}
+
+	hostname, err := valueObject.NewVirtualHost(sslHostname)
+	if err != nil {
+		return ssl, errors.New("SslHostnameError")
+	}
+
+	certInfo, _ := certificate.GetCertInfo()
+
+	certIssuedAtUnix := certInfo.NotBefore.Unix()
+	issuedAt := valueObject.UnixTime(certIssuedAtUnix)
+
+	certExpireAtUnix := certInfo.NotAfter.Unix()
+	expireAt := valueObject.UnixTime(certExpireAtUnix)
+
+	return entity.NewSsl(
+		id,
+		hostname,
+		&issuedAt,
+		&expireAt,
+		certificate,
+		privateKey,
+		chainCertificates,
+	), nil
+}
+
+func (repo SslQueryRepo) GetHttpdVhostsConfig() ([]HttpdVhostConfig, error) {
```

Comment:
Ok, agora entendi a confusão do código ali em cima:

```
	for _, httpdVhostConfig := range httpdVhostsConfig {
		if httpdVhostConfig.VirtualHost != sslToDelete.Hostname.String() {
			continue
		}
```

Você tá precisando de um for loop toda hora por que a sua função que pega o config path do vhost retorna todos os vhosts e todos os paths.

Veja como essa tentativa de economia de tempo gerou um monte de repetição no código com laços desnecessários.

Minha sugestão, separe essas responsabilidades.

Primeiro, crie um método chamado ```GetVirtualHosts() ([]valueObject.Fqdn, error)```. Em seguida, crie um método chamado ```GetVirtualHostConfig(hostname valueObject.Fqdn) (valueObject.UnixFilePath, error)```.

Pronto, agora você não precisa de for loop em lugar nenhum e a sua lógica de extração dos vhosts não precisa se misturar com a lógica de parseamento das chaves. Além de deixar seu código mais aproveitável, vai torná-lo muito mais legível.

Mantenha ambos ainda na SslQueryRepo pois até agora nesse ponto do código, apenas o Ssl precisou desses métodos. Num futuro próximo quando for preciso re-usar esses métodos que você vai criar, mova para uma struct própria e ajuste os SslRepos para usar esse novo método. Ou seja, só estamos separando o código em uma struct própria quando a necessidade surgiu, etapa por etapa.

Daqui pra baixo nem vou revisar o código ainda pois essas sugestões que te dei até agora, já vai mudar drasticamente o código que você produziu até agora.

Reply:
Anotado.

Reply:
Você continuou com a lógica de retornar a listagem de vhosts junto dos filepaths das configs. Essa sua aglutinação que deveria facilitar, nas duas vezes que o GetHttpdVhostsConfig é chamado, você usa um for loop pra ignorar os vhosts (GetVhostConfig) que não precisa ou faz um for loop que ficou difícil de entender (Get do SSL). O código continuou complicado. Não tem por que criar a struct HttpdVhostConfig, ela não tá te ajudando, tá deixando seu código mais difícil de ler.

Reply:
E se eu fizesse uma `map[string]string` e retornasse como chave o `vhost` e o valor `filePath` do `vhost`? Faria mais sentido dessa forma? Porque eu preciso dessas informações para o código funcionar, principalmente porque, como você diss,e eu utilizo em alguns lugares.